compiler/backend/src/org/jetbrains/kotlin/codegen/inline/GlobalInlineContext.kt

```diff
@@ -5,24 +5,48 @@
 
 package org.jetbrains.kotlin.codegen.inline
 
-import org.jetbrains.kotlin.codegen.InlineCycleReporter
+import com.intellij.psi.PsiElement
+import org.jetbrains.kotlin.descriptors.CallableDescriptor
 import org.jetbrains.kotlin.diagnostics.DiagnosticSink
-import org.jetbrains.kotlin.resolve.calls.model.ResolvedCall
+import org.jetbrains.kotlin.diagnostics.Errors
 import java.util.*
 
-class GlobalInlineContext(diagnostics: DiagnosticSink) {
-
-    private val inlineCycleReporter: InlineCycleReporter = InlineCycleReporter(diagnostics)
+class GlobalInlineContext(private val diagnostics: DiagnosticSink) {
+    // Ordered set of declarations and inline calls being generated right now.
+    // No call in it should point to a declaration that's before it in the stack.
+    private val inlineCallsAndDeclarations = LinkedList<Any? /* CallableDescriptor | PsiElement? */>()
+    private val inlineDeclarationSet = mutableSetOf<CallableDescriptor>()
 
     private val typesUsedInInlineFunctions = LinkedList<MutableSet<String>>()
 
-    fun enterIntoInlining(call: InlineCall?) =
-        inlineCycleReporter.enterIntoInlining(call).also {
-            if (it) typesUsedInInlineFunctions.push(hashSetOf())
+    fun enterDeclaration(descriptor: CallableDescriptor) {
```

Comment:
What about inline flag check?

Reply:
It's not really necessary, all other declarations are simply not used for anything.